lib/Sema/TypeCheckDeclPrimary.cpp

```diff
@@ -3351,6 +3635,10 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
       (void) CD->getDestructor();
       
       // FIXME: Should we duplicate any other logic from visitClassDecl()?
+
+      evaluateOrDefault(Ctx.evaluator,
```

Comment:
In a multi-file module with `@_objcImplementation` extensions scattered across multiple files, will we see duplicated diagnostics?

Reply:
I think you would. It might be better to match one extension with its interface at a time so we only diagnose problems with extensions in a primary file; I didn't do that because it made it easier to give good you-put-this-in-the-wrong-extension diagnostics (while being aware of which decls had already been matched to something else), but maybe I should accept those possibly being a little worse.

Reply:
The upcoming push diagnoses one extension and its interface at a time, instead of a whole class. That regresses the quality of the diagnostics when an implementation is in the wrong extension—you get a "missing implementation" error in one extension and an "unmatched implementation" error in the other, instead of a tailored "hey, this should have been over there" diagnostic we emitted before—but it ensures we only check and diagnose extensions in a primary file. It also means that the checker is evaluating much smaller sets of members at a time, which I expect will improve performance since this is a very O(n<sup>2</sup>) sort of operation.